api/src/main/java/org/apache/iceberg/PartitionSpec.java

```diff
@@ -319,7 +319,7 @@ public static Builder builderFor(Schema schema) {
     private final Schema schema;
     private final List<PartitionField> fields = Lists.newArrayList();
     private final Set<String> partitionNames = Sets.newHashSet();
-    private Map<Integer, PartitionField> timeFields = Maps.newHashMap();
+    private Map<String, PartitionField> partitionFields = Maps.newHashMap();
```

Comment:
Overall LGTM, one nit that I think `partitionFields` here would be good to be renamed so that it's easy to tell it's just for collision detecting. Also I wonder if we want to do this collision check for other transformations too (like bucket, and record numBuckets in the string), so that we might be able to combine `fields` and `partitionFields` into potentially something like a LinkedHashMap?

Reply:
+1, maybe some name like a `dedupPartitionFields`. And if we are doing this for datetime transforms, we should probably also do it for bucket and truncate and alwaysNull.

Reply:
Thanks @yyanyy and @jackye1995 for the review.
`alwaysNull` should not be deduped as the same source field might be removed multiple times during the spec evolution.

Reply:
I like this implementation because it allows easily adding new checks. I agree that the name could be better, like `dedupFields` or `dedupPartitionFields` as Jack suggested. I would not combine this with `fields`. I think that would make both too difficult to reason about.

One more thing I would change is to make the key a `Pair<Integer, String>` instead of a `String` That avoids confusion about how to create the keys, and also prevents unnecessary collisions like `(1, p1)` with `(11, p)`.

Reply:
Updated. As api module cannot use Pair from core module, I used map entry for the same purpose.